File: mycelium/languages/csharp.py
"""C# language analyser."""

from __future__ import annotations

import tree_sitter
import tree_sitter_c_sharp as ts_csharp

from mycelium.config import ImportStatement, RawCall, Symbol, SymbolType, Visibility

# Map C# AST node types to our SymbolType
_TYPE_MAP = {
    "class_declaration": SymbolType.CLASS,
    "interface_declaration": SymbolType.INTERFACE,
    "struct_declaration": SymbolType.STRUCT,
    "enum_declaration": SymbolType.ENUM,
    "method_declaration": SymbolType.METHOD,
    "constructor_declaration": SymbolType.CONSTRUCTOR,
    "property_declaration": SymbolType.PROPERTY,
    "namespace_declaration": SymbolType.NAMESPACE,
    "record_declaration": SymbolType.RECORD,
    "delegate_declaration": SymbolType.DELEGATE,
}

# C# modifier keywords to Visibility
_VISIBILITY_MAP = {
    "public": Visibility.PUBLIC,
    "private": Visibility.PRIVATE,
    "internal": Visibility.INTERNAL,
    "protected": Visibility.PROTECTED,
}

# Node types that can contain child symbols
_CONTAINER_TYPES = {
    "class_declaration", "interface_declaration", "struct_declaration",
    "record_declaration", "namespace_declaration",
}


class CSharpAnalyser:
    extensions = [".cs"]
    language_name = "cs"

    def get_language(self) -> tree_sitter.Language:
        return tree_sitter.Language(ts_csharp.language())

    def extract_symbols(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[Symbol]:
        symbols: list[Symbol] = []
        self._walk_node(tree.root_node, source, file_path, symbols, parent_id=None)
        return symbols

    def _walk_node(
        self,
        node: tree_sitter.Node,
        source: bytes,
        file_path: str,
        symbols: list[Symbol],
        parent_id: str | None,
    ) -> None:
        """Recursively walk AST nodes to extract symbol definitions."""
        for child in node.children:
            symbol_type = _TYPE_MAP.get(child.type)
            if symbol_type is not None:
                name = self._get_name(child)
                if name is None:
                    continue

                visibility = self._get_visibility(child)

                # Namespaces don't have visibility modifiers in C#
                if symbol_type == SymbolType.NAMESPACE:
                    visibility = Visibility.UNKNOWN

                exported = visibility in (Visibility.PUBLIC, Visibility.INTERNAL)

                # Use a placeholder ID - will be replaced by parsing phase
                sym_id = f"_pending_{len(symbols)}"
                # Extract constructor parameter types for DI resolution
                param_types = None
                if child.type == "constructor_declaration":
                    param_types = self._extract_parameter_types(child)

                sym = Symbol(
                    id=sym_id,
                    name=name,
                    type=symbol_type,
                    file=file_path,
                    line=child.start_point[0] + 1,
                    visibility=visibility,
                    exported=exported,
                    parent=parent_id,
                    byte_range=(child.start_byte, child.end_byte),
                    parameter_types=param_types,
                )
                symbols.append(sym)

                # Recurse into containers for nested symbols
                if child.type in _CONTAINER_TYPES:
                    decl_list = child.child_by_field_name("body")
                    if decl_list is None:
                        # Try declaration_list for namespaces/classes
                        for c in child.children:
                            if c.type == "declaration_list":
                                decl_list = c
                                break
                    if decl_list:
                        self._walk_node(decl_list, source, file_path, symbols, parent_id=name)
            elif child.type == "file_scoped_namespace_declaration":
                # Handle file-scoped namespaces (C# 10+): namespace MyApp;
                name = self._get_name(child)
                if name:
                    sym = Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.NAMESPACE,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.UNKNOWN,
                        exported=True,
                        parent=parent_id,
                    )
                    symbols.append(sym)
                    # File-scoped namespace contains the rest of the file
                    self._walk_node(child, source, file_path, symbols, parent_id=name)

    def _get_name(self, node: tree_sitter.Node) -> str | None:
        """Extract the identifier name from a declaration node."""
        # Prefer the 'name' field - handles methods returning custom types
        # where the first identifier is the return type, not the name
        name_node = node.child_by_field_name("name")
        if name_node:
            return name_node.text.decode("utf-8")
        # Fallback for constructors and other nodes without a name field
        for child in node.children:
            if child.type == "identifier":
                return child.text.decode("utf-8")
            if child.type == "qualified_name":
                return child.text.decode("utf-8")
        return None

    def _get_visibility(self, node: tree_sitter.Node) -> Visibility:
        """Extract visibility from modifier nodes."""
        for child in node.children:
            if child.type == "modifier":
                mod_text = child.text.decode("utf-8").lower()
                vis = _VISIBILITY_MAP.get(mod_text)
                if vis:
                    return vis
        return Visibility.PRIVATE  # C# default is private

    def extract_imports(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[ImportStatement]:
        imports: list[ImportStatement] = []
        for child in tree.root_node.children:
            if child.type == "using_directive":
                # Extract the namespace from the using directive
                name_node = None
                for c in child.children:
                    if c.type in ("identifier", "qualified_name"):
                        name_node = c
                        break
                    if c.type == "name":
                        name_node = c
                        break
                if name_node:
                    target = name_node.text.decode("utf-8")
                    statement = child.text.decode("utf-8").rstrip(";").strip()
                    imports.append(ImportStatement(
                        file=file_path,
                        statement=statement,
                        target_name=target,
                        line=child.start_point[0] + 1,
                    ))
            elif child.type in ("namespace_declaration", "file_scoped_namespace_declaration"):
                # Check for using directives inside namespace
                for ns_child in child.children:
                    if ns_child.type == "declaration_list":
                        for decl_child in ns_child.children:
                            if decl_child.type == "using_directive":
                                name_node = None
                                for c in decl_child.children:
                                    if c.type in ("identifier", "qualified_name"):
                                        name_node = c
                                        break
                                if name_node:
                                    target = name_node.text.decode("utf-8")
                                    statement = decl_child.text.decode("utf-8").rstrip(";").strip()
                                    imports.append(ImportStatement(
                                        file=file_path,
                                        statement=statement,
                                        target_name=target,
                                        line=decl_child.start_point[0] + 1,
                                    ))
        return imports

    def extract_calls(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[RawCall]:
        calls: list[RawCall] = []
        exclusions = self.builtin_exclusions()
        self._find_calls(tree.root_node, source, file_path, calls, exclusions)
        return calls

    def _find_calls(
        self,
        node: tree_sitter.Node,
        source: bytes,
        file_path: str,
        calls: list[RawCall],
        exclusions: set[str],
    ) -> None:
        """Recursively find call expressions in the AST."""
        if node.type == "invocation_expression":
            callee_name, qualifier = self._extract_callee(node)
            if callee_name and callee_name not in exclusions:
                qualified = f"{qualifier}.{callee_name}" if qualifier else callee_name
                if qualified not in exclusions:
                    caller = self._find_enclosing_method(node)
                    calls.append(RawCall(
                        caller_file=file_path,
                        caller_name=caller or "<module>",
                        callee_name=callee_name,
                        line=node.start_point[0] + 1,
                        qualifier=qualifier,
                    ))
        elif node.type == "object_creation_expression":
            # new SomeClass() is a call to the constructor
            callee_name = None
            for child in node.children:
                if child.type == "identifier":
                    callee_name = child.text.decode("utf-8")
                    break
                if child.type == "qualified_name":
                    callee_name = child.text.decode("utf-8")
                    break
            if callee_name and callee_name not in exclusions:
                caller = self._find_enclosing_method(node)
                calls.append(RawCall(
                    caller_file=file_path,
                    caller_name=caller or "<module>",
                    callee_name=callee_name,
                    line=node.start_point[0] + 1,
                    qualifier=None,
                ))

        for child in node.children:
            self._find_calls(child, source, file_path, calls, exclusions)

    def _extract_callee(self, inv_node: tree_sitter.Node) -> tuple[str | None, str | None]:
        """Extract callee name and qualifier from an invocation_expression."""
        first_child = inv_node.children[0] if inv_node.children else None
        if first_child is None:
            return None, None

        if first_child.type == "identifier":
            return first_child.text.decode("utf-8"), None
        elif first_child.type == "member_access_expression":
            # e.g., _service.CalculateEntitlement or Console.WriteLine
            parts = []
            for child in first_child.children:
                if child.type == "identifier":
                    parts.append(child.text.decode("utf-8"))
            if len(parts) >= 2:
                return parts[-1], parts[-2]
            elif len(parts) == 1:
                return parts[0], None
        elif first_child.type == "qualified_name":
            text = first_child.text.decode("utf-8")
            parts = text.rsplit(".", 1)
            if len(parts) == 2:
                return parts[1], parts[0]
            return text, None

        return None, None

    def _find_enclosing_method(self, node: tree_sitter.Node) -> str | None:
        """Walk up the AST to find the enclosing method or constructor.

        Stops at property/event/operator declarations to prevent attributing
        calls in accessors to the wrong symbol.
        """
        current = node.parent
        while current:
            if current.type in ("method_declaration", "constructor_declaration",
                                "local_function_statement"):
                name_node = current.child_by_field_name("name")
                if name_node:
                    return name_node.text.decode("utf-8")
                for child in current.children:
                    if child.type == "identifier":
                        return child.text.decode("utf-8")
            # Stop at property/event/operator boundaries - calls inside
            # these should not be attributed to an outer method
            if current.type in ("property_declaration", "event_declaration",
                                "operator_declaration", "indexer_declaration"):
                return None
            current = current.parent
        return None

    def _extract_parameter_types(self, node: tree_sitter.Node) -> list[tuple[str, str]] | None:
        """Extract (param_name, type_name) pairs from a constructor's parameter_list."""
        param_list = node.child_by_field_name("parameters")
        if param_list is None:
            return None
        params = []
        for child in param_list.children:
            if child.type == "parameter":
                type_node = child.child_by_field_name("type")
                name_node = child.child_by_field_name("name")
                if type_node and name_node:
                    type_name = type_node.text.decode("utf-8")
                    param_name = name_node.text.decode("utf-8")
                    params.append((param_name, type_name))
        return params if params else None

    def builtin_exclusions(self) -> set[str]:
        return {
            # Framework types (not real calls)
            "Task", "ValueTask",
            # Console
            "Console.WriteLine", "Console.ReadLine", "Console.Write",
            "Console.ReadKey", "Console.Clear",
            # String
            "String.Format", "String.IsNullOrEmpty", "String.IsNullOrWhiteSpace",
            "String.Join", "String.Concat", "String.Compare",
            "string.Format", "string.IsNullOrEmpty", "string.IsNullOrWhiteSpace",
            "string.Join", "string.Concat", "string.Compare",
            # Convert
            "Convert.ToInt32", "Convert.ToString", "Convert.ToDecimal",
            "Convert.ToDouble", "Convert.ToBoolean", "Convert.ToDateTime",
            # Math
            "Math.Abs", "Math.Max", "Math.Min", "Math.Round",
            "Math.Floor", "Math.Ceiling", "Math.Pow", "Math.Sqrt",
            # Object
            "ToString", "Equals", "GetHashCode", "GetType",
            # Debug/Trace
            "Debug.WriteLine", "Debug.Assert", "Debug.Print",
            "Trace.WriteLine", "Trace.TraceInformation",
            # GC
            "GC.Collect", "GC.SuppressFinalize",
            # Task
            "Task.Run", "Task.WhenAll", "Task.WhenAny", "Task.Delay",
            "Task.FromResult", "Task.CompletedTask",
            # LINQ
            "Select", "Where", "FirstOrDefault", "First", "Last",
            "LastOrDefault", "SingleOrDefault", "Single", "Any", "All",
            "Count", "Sum", "Average", "Min", "Max", "OrderBy",
            "OrderByDescending", "GroupBy", "ToList", "ToArray",
            "ToDictionary", "AsEnumerable", "AsQueryable",
            "Skip", "Take", "Distinct", "Union", "Intersect", "Except",
            "Aggregate", "Zip", "SelectMany", "Contains",
            # Common framework
            "Dispose", "Close",
        }

File: crates/mycelium-core/src/languages/csharp.rs
//! C# language analyser.

use std::collections::HashSet;
use std::sync::LazyLock;

use tree_sitter::{Language, Node, Tree};

use super::LanguageAnalyser;
use crate::config::{ImportStatement, RawCall, Symbol, SymbolType, Visibility};

static BUILTIN_EXCLUSIONS: LazyLock<HashSet<String>> = LazyLock::new(|| {
    [
        // Framework types
        "Task",
        "ValueTask",
        // Console
        "Console.WriteLine",
        "Console.ReadLine",
        "Console.Write",
        "Console.ReadKey",
        "Console.Clear",
        // String
        "String.Format",
        "String.IsNullOrEmpty",
        "String.IsNullOrWhiteSpace",
        "String.Join",
        "String.Concat",
        "String.Compare",
        "string.Format",
        "string.IsNullOrEmpty",
        "string.IsNullOrWhiteSpace",
        "string.Join",
        "string.Concat",
        "string.Compare",
        // Convert
        "Convert.ToInt32",
        "Convert.ToString",
        "Convert.ToDecimal",
        "Convert.ToDouble",
        "Convert.ToBoolean",
        "Convert.ToDateTime",
        // Math
        "Math.Abs",
        "Math.Max",
        "Math.Min",
        "Math.Round",
        "Math.Floor",
        "Math.Ceiling",
        "Math.Pow",
        "Math.Sqrt",
        // Object
        "ToString",
        "Equals",
        "GetHashCode",
        "GetType",
        "ReferenceEquals",
        "MemberwiseClone",
        "Finalize",
        // Debug/Trace
        "Debug.WriteLine",
        "Debug.Write",
        "Debug.Assert",
        "Debug.Print",
        "Trace.WriteLine",
        "Trace.TraceInformation",
        // GC
        "GC.Collect",
        "GC.SuppressFinalize",
        // Task
        "Task.Run",
        "Task.WhenAll",
        "Task.WhenAny",
        "Task.Delay",
        "Task.FromResult",
        "Task.CompletedTask",
        "ValueTask.FromResult",
        "ValueTask.CompletedTask",
        // Parsing
        "int.Parse",
        "int.TryParse",
        "Guid.NewGuid",
        "Guid.Parse",
        "Guid.TryParse",
        // Keywords
        "nameof",
        "typeof",
        "sizeof",
        "default",
        "ArgumentNullException.ThrowIfNull",
        // LINQ
        "Select",
        "Where",
        "FirstOrDefault",
        "First",
        "Last",
        "LastOrDefault",
        "SingleOrDefault",
        "Single",
        "Any",
        "All",
        "Count",
        "Sum",
        "Average",
        "Min",
        "Max",
        "OrderBy",
        "OrderByDescending",
        "GroupBy",
        "ToList",
        "ToArray",
        "ToDictionary",
        "AsEnumerable",
        "AsQueryable",
        "Skip",
        "Take",
        "Distinct",
        "Union",
        "Intersect",
        "Except",
        "Aggregate",
        "Zip",
        "SelectMany",
        "Contains",
        // Common framework
        "Dispose",
        "Close",
    ]
    .into_iter()
    .map(String::from)
    .collect()
});

fn node_to_symbol_type(node_type: &str) -> Option<SymbolType> {
    match node_type {
        "class_declaration" => Some(SymbolType::Class),
        "interface_declaration" => Some(SymbolType::Interface),
        "struct_declaration" => Some(SymbolType::Struct),
        "enum_declaration" => Some(SymbolType::Enum),
        "namespace_declaration" | "file_scoped_namespace_declaration" => {
            Some(SymbolType::Namespace)
        }
        "record_declaration" => Some(SymbolType::Record),
        "delegate_declaration" => Some(SymbolType::Delegate),
        "method_declaration" => Some(SymbolType::Method),
        "constructor_declaration" => Some(SymbolType::Constructor),
        "property_declaration" => Some(SymbolType::Property),
        _ => None,
    }
}

fn is_container(node_type: &str) -> bool {
    matches!(
        node_type,
        "class_declaration"
            | "struct_declaration"
            | "interface_declaration"
            | "record_declaration"
            | "namespace_declaration"
            | "file_scoped_namespace_declaration"
    )
}

fn get_visibility(node: &Node, source: &[u8]) -> Visibility {
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if child.kind() == "modifier" {
                let mod_text = child.utf8_text(source).unwrap_or("").to_lowercase();
                match mod_text.as_str() {
                    "public" => return Visibility::Public,
                    "private" => return Visibility::Private,
                    "internal" => return Visibility::Internal,
                    "protected" => return Visibility::Protected,
                    _ => {}
                }
            }
        }
    }
    Visibility::Private // C# default
}

fn get_name(node: &Node, source: &[u8]) -> Option<String> {
    if let Some(name_node) = node.child_by_field_name("name") {
        return name_node.utf8_text(source).ok().map(|s| s.to_string());
    }
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if child.kind() == "identifier" {
                return child.utf8_text(source).ok().map(|s| s.to_string());
            }
        }
    }
    None
}

pub struct CSharpAnalyser;

impl Default for CSharpAnalyser {
    fn default() -> Self {
        Self
    }
}

impl CSharpAnalyser {
    pub fn new() -> Self {
        Self
    }

    fn walk_node(
        &self,
        node: &Node,
        source: &[u8],
        file_path: &str,
        symbols: &mut Vec<Symbol>,
        parent_id: Option<&str>,
    ) {
        for i in 0..node.child_count() {
            let child = match node.child(i) {
                Some(c) => c,
                None => continue,
            };

            let sym_type = match node_to_symbol_type(child.kind()) {
                Some(t) => t,
                None => continue,
            };

            let name = match get_name(&child, source) {
                Some(n) => n,
                None => continue,
            };

            let mut vis = get_visibility(&child, source);
            let mut exported = vis == Visibility::Public;

            // Namespaces have no visibility modifiers
            if sym_type == SymbolType::Namespace {
                vis = Visibility::Unknown;
                exported = true;
            }

            // Extract parameter types for constructors (DI tracking)
            let parameter_types = if child.kind() == "constructor_declaration" {
                extract_parameter_types(&child, source)
            } else {
                None
            };

            symbols.push(Symbol {
                id: format!("_pending_{}", symbols.len()),
                name: name.clone(),
                symbol_type: sym_type,
                file: file_path.to_string(),
                line: child.start_position().row + 1,
                visibility: vis,
                exported,
                parent: parent_id.map(|s| s.to_string()),
                language: Some("C#".to_string()),
                byte_range: Some((child.byte_range().start, child.byte_range().end)),
                parameter_types,
            });

            // Recurse into containers
            if is_container(child.kind()) {
                for j in 0..child.child_count() {
                    if let Some(c) = child.child(j) {
                        if c.kind() == "declaration_list" {
                            self.walk_node(&c, source, file_path, symbols, Some(&name));
                            break;
                        }
                    }
                }
            }
        }
    }

    fn extract_using(
        &self,
        node: &Node,
        source: &[u8],
        file_path: &str,
    ) -> Option<ImportStatement> {
        let mut name_node = None;
        for i in 0..node.child_count() {
            if let Some(child) = node.child(i) {
                match child.kind() {
                    "identifier" | "qualified_name" | "name" => {
                        name_node = Some(child);
                        break;
                    }
                    _ => {}
                }
            }
        }
        let name_node = name_node?;
        let target = name_node.utf8_text(source).ok()?.to_string();
        let statement = node
            .utf8_text(source)
            .ok()?
            .trim_end_matches(';')
            .trim()
            .to_string();
        Some(ImportStatement {
            file: file_path.to_string(),
            statement,
            target_name: target,
            line: node.start_position().row + 1,
        })
    }

    fn find_calls(
        &self,
        node: &Node,
        source: &[u8],
        file_path: &str,
        calls: &mut Vec<RawCall>,
        exclusions: &HashSet<String>,
    ) {
        if node.kind() == "invocation_expression" {
            let (callee_name, qualifier) = extract_callee(node, source);
            if let Some(ref name) = callee_name {
                if !exclusions.contains(name) {
                    let qualified = if let Some(ref q) = qualifier {
                        format!("{}.{}", q, name)
                    } else {
                        name.clone()
                    };
                    if !exclusions.contains(&qualified) {
                        let caller = find_enclosing_method(node, source);
                        calls.push(RawCall {
                            caller_file: file_path.to_string(),
                            caller_name: caller.unwrap_or_else(|| "<module>".to_string()),
                            callee_name: name.clone(),
                            line: node.start_position().row + 1,
                            qualifier,
                        });
                    }
                }
            }
        } else if node.kind() == "object_creation_expression" {
            let mut callee_name = None;
            for i in 0..node.child_count() {
                if let Some(child) = node.child(i) {
                    if child.kind() == "identifier" || child.kind() == "qualified_name" {
                        callee_name = child.utf8_text(source).ok().map(|s| s.to_string());
                        break;
                    }
                }
            }
            if let Some(ref name) = callee_name {
                if !exclusions.contains(name) {
                    let caller = find_enclosing_method(node, source);
                    calls.push(RawCall {
                        caller_file: file_path.to_string(),
                        caller_name: caller.unwrap_or_else(|| "<module>".to_string()),
                        callee_name: name.clone(),
                        line: node.start_position().row + 1,
                        qualifier: None,
                    });
                }
            }
        }

        for i in 0..node.child_count() {
            if let Some(child) = node.child(i) {
                self.find_calls(&child, source, file_path, calls, exclusions);
            }
        }
    }
}

fn extract_parameter_types(node: &Node, source: &[u8]) -> Option<Vec<(String, String)>> {
    let param_list = node.child_by_field_name("parameters")?;
    let mut params = Vec::new();
    for i in 0..param_list.child_count() {
        if let Some(child) = param_list.child(i) {
            if child.kind() == "parameter" {
                let type_node = child.child_by_field_name("type");
                let name_node = child.child_by_field_name("name");
                if let (Some(tn), Some(nn)) = (type_node, name_node) {
                    if let (Ok(type_name), Ok(param_name)) =
                        (tn.utf8_text(source), nn.utf8_text(source))
                    {
                        params.push((param_name.to_string(), type_name.to_string()));
                    }
                }
            }
        }
    }
    if params.is_empty() {
        None
    } else {
        Some(params)
    }
}

fn extract_callee(inv_node: &Node, source: &[u8]) -> (Option<String>, Option<String>) {
    let first_child = match inv_node.child(0) {
        Some(c) => c,
        None => return (None, None),
    };

    match first_child.kind() {
        "identifier" => {
            let name = first_child.utf8_text(source).ok().map(|s| s.to_string());
            (name, None)
        }
        "member_access_expression" => {
            let mut parts = Vec::new();
            for i in 0..first_child.child_count() {
                if let Some(child) = first_child.child(i) {
                    if child.kind() == "identifier" {
                        if let Ok(text) = child.utf8_text(source) {
                            parts.push(text.to_string());
                        }
                    }
                }
            }
            if parts.len() >= 2 {
                let name = parts.pop();
                let qualifier = parts.pop();
                (name, qualifier)
            } else if parts.len() == 1 {
                (Some(parts.remove(0)), None)
            } else {
                (None, None)
            }
        }
        "qualified_name" => {
            let text = first_child.utf8_text(source).unwrap_or("");
            if let Some(pos) = text.rfind('.') {
                (
                    Some(text[pos + 1..].to_string()),
                    Some(text[..pos].to_string()),
                )
            } else {
                (Some(text.to_string()), None)
            }
        }
        _ => (None, None),
    }
}

fn find_enclosing_method(node: &Node, source: &[u8]) -> Option<String> {
    let mut current = node.parent();
    while let Some(n) = current {
        match n.kind() {
            "method_declaration" | "constructor_declaration" | "local_function_statement" => {
                if let Some(name_node) = n.child_by_field_name("name") {
                    return name_node.utf8_text(source).ok().map(|s| s.to_string());
                }
                for i in 0..n.child_count() {
                    if let Some(child) = n.child(i) {
                        if child.kind() == "identifier" {
                            return child.utf8_text(source).ok().map(|s| s.to_string());
                        }
                    }
                }
            }
            // Stop at property/event/operator boundaries
            "property_declaration"
            | "event_declaration"
            | "operator_declaration"
            | "indexer_declaration" => {
                return None;
            }
            _ => {}
        }
        current = n.parent();
    }
    None
}

impl LanguageAnalyser for CSharpAnalyser {
    fn extensions(&self) -> &[&str] {
        &["cs"]
    }

    fn language_name(&self) -> &str {
        "C#"
    }

    fn get_language(&self) -> Language {
        tree_sitter_c_sharp::LANGUAGE.into()
    }

    fn extract_symbols(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<Symbol> {
        let mut symbols = Vec::new();
        self.walk_node(&tree.root_node(), source, file_path, &mut symbols, None);
        symbols
    }

    fn extract_imports(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<ImportStatement> {
        let mut imports = Vec::new();
        let root = tree.root_node();
        for i in 0..root.child_count() {
            if let Some(child) = root.child(i) {
                if child.kind() == "using_directive" {
                    if let Some(imp) = self.extract_using(&child, source, file_path) {
                        imports.push(imp);
                    }
                } else if child.kind() == "namespace_declaration"
                    || child.kind() == "file_scoped_namespace_declaration"
                {
                    // Check for using directives inside namespace
                    for j in 0..child.child_count() {
                        if let Some(ns_child) = child.child(j) {
                            if ns_child.kind() == "declaration_list" {
                                for k in 0..ns_child.child_count() {
                                    if let Some(decl_child) = ns_child.child(k) {
                                        if decl_child.kind() == "using_directive" {
                                            if let Some(imp) =
                                                self.extract_using(&decl_child, source, file_path)
                                            {
                                                imports.push(imp);
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        imports
    }

    fn extract_calls(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<RawCall> {
        let mut calls = Vec::new();
        let exclusions = self.builtin_exclusions();
        self.find_calls(&tree.root_node(), source, file_path, &mut calls, exclusions);
        calls
    }

    fn builtin_exclusions(&self) -> &HashSet<String> {
        &BUILTIN_EXCLUSIONS
    }
}
